python_modules/dagster/dagster/_core/execution/context/invocation.py

```diff
@@ -129,6 +150,114 @@ def __exit__(self, *exc):
+            if invocation_tags
+            else self._op_def.tags
+        )
+
+        self._hook_defs = (
+            pending_invocation.hook_defs
+            if isinstance(pending_invocation, PendingNodeInvocation)
+            else None
+        )
+        invocation_alias = (
+            pending_invocation.given_alias
+            if isinstance(pending_invocation, PendingNodeInvocation)
+            else None
+        )
+        self._alias = invocation_alias if invocation_alias else self._op_def.name
+
+        self._assets_def = assets_def
+
+        if resources_from_args:
+            if self._resource_defs:
+                raise DagsterInvalidInvocationError(
+                    "Cannot provide resources in both context and kwargs"
+                )
+            resource_defs = wrap_resources_for_execution(resources_from_args)
+            # add new resources context to the stack to be cleared on exit
+            self._resources = self._exit_stack.enter_context(
+                build_resources(resource_defs, self.instance)
+            )
+        elif assets_def and assets_def.resource_defs:
+            for key in sorted(list(assets_def.resource_defs.keys())):
+                if key in self._resource_defs:
+                    raise DagsterInvalidInvocationError(
+                        f"Error when invoking {assets_def!s} resource '{key}' "
+                        "provided on both the definition and invocation context. Please "
+                        "provide on only one or the other."
+                    )
+            resource_defs = wrap_resources_for_execution(
+                {**self._resource_defs, **assets_def.resource_defs}
+            )
+            # add new resources context to the stack to be cleared on exit
+            self._resources = self._exit_stack.enter_context(
+                build_resources(resource_defs, self.instance, self._resources_config)
+            )
+        else:
+            # this runs the check in resources() to ensure we are in a context manager if necessary
+            self._resources = self.resources
+
+            resource_defs = self._resource_defs
+
+        _validate_resource_requirements(resource_defs, op_def)
+
+        if self.op_config and config_from_args:
+            raise DagsterInvalidInvocationError("Cannot provide config in both context and kwargs")
+        self._op_config = resolve_bound_config(config_from_args or self.op_config, op_def)
+
+        self._requires_typed_event_stream = False
+        self._typed_event_stream_error_message = None
+        self._bound = True
+
+        return self
+
+    def unbind(self):
+        self._op_def = None
+        self._alias = None
+        self._hook_defs = None
+        self._tags = {}
+        self._seen_outputs = {}
+        self._resources = self._init_resources
+        self._op_config = self._init_op_config
+        self._user_events = []
+        self._output_metadata = {}
```

Comment:
just continuing to refine the thinking here, theres like 3 types of information on this object 
* properties that were set at context creation 
* properties determined by what the context was used on and with to invoke ("bind time") 
* properties that were resolved during invocation 

I can imagine bundling the bind time arguments together as mentioned above 

formalizing the during/post invocation state may help with [2]

Reply:
to make sure I'm clear on the distinction between the last two - are you thinking the "properties determined by what the context was used on and with to invoke ("bind time")" is basically everything that could be determined in the `bind` function and "properties that were resolved during invocation" are things like events and output metadata?

Reply:
yep that sounds correct to me